File: time_diff/time_diff.py
import time
from datetime import datetime

def format_timestamp(ts):
    dt = datetime.fromtimestamp(ts)
    ns_total = int((ts - int(ts)) * 1_000_000_000)
    ms = ns_total // 1_000_000
    us = (ns_total // 1_000) % 1_000
    ns_last_2 = (ns_total // 10) % 100

    return dt.strftime(f"%Y-%m-%d_%H%M_%S_{ms:03d}_{us:03d}_{ns_last_2:02d}")

# Start
start = time.time_ns()
start_sec = start / 1_000_000_000
start_fmt = format_timestamp(start_sec)

time.sleep(5)

# End
end = time.time_ns()
end_sec = end / 1_000_000_000
end_fmt = format_timestamp(end_sec)

# Difference
delta_ns = end - start
delta_sec = delta_ns / 1_000_000_000
delta_ms = (delta_ns // 1_000_000) % 1000
delta_us = (delta_ns // 1_000) % 1000
delta_ns_last_2 = (delta_ns // 10) % 100
delta_int_sec = delta_ns // 1_000_000_000

# Output
print("Start formatted :", start_fmt)
print("End formatted   :", end_fmt)
print()

print("⏱ Full Range Time Difference:")
print(f"-> Human Format Delta : 0000-00-00_0000_{delta_int_sec:02d}_{delta_ms:03d}_{delta_us:03d}_{delta_ns_last_2:02d}")
print(f"-> Epoch Format Delta : {delta_sec:.9f} seconds")

File: time_diff/time_diff.go
package main

import (
	"fmt"
	"time"
)

func formatTimestamp(t time.Time) string {
	return fmt.Sprintf("%04d-%02d-%02d_%02d%02d_%02d_%03d_%03d_%02d",
		t.Year(), t.Month(), t.Day(),
		t.Hour(), t.Minute(), t.Second(),
		t.Nanosecond()/1e6,      // milliseconds
		(t.Nanosecond()/1e3)%1e3, // microseconds
		(t.Nanosecond()/1e1)%1e2, // last 2 digits of nanoseconds
	)
}

func main() {
	start := time.Now()
	startFmt := formatTimestamp(start)

	time.Sleep(5 * time.Second)

	end := time.Now()
	endFmt := formatTimestamp(end)

	duration := end.Sub(start)
	seconds := duration.Seconds()
	nanos := duration.Nanoseconds()

	ms := (nanos / 1e6)
	us := (nanos / 1e3) % 1000
	ns := (nanos / 1e1) % 100

	fmt.Println("Start time formatted :", startFmt)
	fmt.Println("End time formatted   :", endFmt)
	fmt.Println()

	fmt.Printf("⏱ Full Range Time Difference:\n")
	fmt.Printf("-> Human Format Delta : 0000-00-00_0000_%02d_%03d_%03d_%02d\n",
		int(duration.Seconds()), ms%1000, us, ns)
	fmt.Printf("-> Epoch Format Delta : %.9f seconds\n", seconds)
}


/*
go mod tidy
go run time_diff.go

// build it as a standalone binary:
go build -o time_diff_go time_diff.go
./time_diff_go

*/